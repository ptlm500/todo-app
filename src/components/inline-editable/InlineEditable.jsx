import React, { useRef, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

export default function InlineEditable(props) {
  const {
    className,
    value,
    placeholder,
    children,
    onSubmit
  } = props;

  if (Array.isArray(children)) {
    throw new Error('[InlineEditable]: Multiple child components passed. InlineEditable only accepts one child', children);
  }

  const childRef = useRef();

  const [editing, setEditing] = useState(false);

  useEffect(() => {
    if (editing && childRef && childRef.current) {
      childRef.current.focus();
    }
  }, [editing, childRef]);

  const childWithProps = React.cloneElement(children, {
    ref: childRef,
    value,
    placeholder
  });

  function handleKeyDown(e) {
    if (e.key === 'Escape') {
      finishEdit();
    }
  }

  function finishEdit() {
    setEditing(false);
    onSubmit && onSubmit(value);
  }

  return (
    <div
      className={className ? `${className} inline-editable` : 'inline-editable'}
    >
      {
        editing ? (
          <div
            className='inline-editable__edit-container'
            onBlur={finishEdit}
            onKeyDown={handleKeyDown}
          >
            {childWithProps}
          </div>
        ) : (
          <div
            className='inline-editable__edit-trigger'
            onClick={() => setEditing(true)}
          >
            <Value>
              {value || placeholder || 'Provide a placeholder value'}
            </Value>
          </div>
        )
      }
    </div>
  );
}

const Value = styled.span`
  white-space: pre-wrap;
  word-break: break-word;
`;

InlineEditable.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
  value: PropTypes.string,
  placeholder: PropTypes.string,
  onSubmit: PropTypes.func
};
