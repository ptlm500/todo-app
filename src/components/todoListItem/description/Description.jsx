import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import InlineEditable from '../../inline-editable';
import AutosizingTextarea from '../../autosizingTextarea';
import { theme } from '../../../styles';

function Description(props) {
  const {
    className,
    updateDescription
  } = props;
  const descriptionInStore = props.description;

  const [description, setDescription] = useState(descriptionInStore);

  useEffect(() => {
    // Set description state if stored description changes
    setDescription(descriptionInStore);
  }, [descriptionInStore]);

  return (
    <InlineEditable
      className={className}
      value={description}
      placeholder={' '}
      onSubmit={updateDescription}
    >
      <AutosizingTextarea
        onChange={e => setDescription(e.target.value)}
      />
    </InlineEditable>
  );
}

Description.propTypes = {
  className: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  updateDescription: PropTypes.func.isRequired
};

export default styled(Description)`
  ${theme.bodyFontSize};
  margin-bottom: 1rem;
  min-width: 100px;
`;
