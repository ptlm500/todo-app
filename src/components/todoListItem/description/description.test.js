import React from 'react';
import renderer, { act } from 'react-test-renderer';
import Description from './Description';
import InlineEditable from '../../inline-editable';

const defaultProps = {
  description: 'Test description',
  updateDescription: jest.fn(),
  className: 'test-class'
};

beforeEach(() => {
  jest.clearAllMocks();
});

describe('Description', () => {
  it('renders a Description', () => {
    let tree;
    act(() => {
      tree = renderer.create(<Description {...defaultProps}/>);
    });

    const inlineEditable = tree.root.findByType(InlineEditable);
    expect(inlineEditable.props.value).toBe(defaultProps.description);
    expect(inlineEditable.props.placeholder).toBe(' ');

    const newDescription = 'New description';

    inlineEditable.props.onSubmit(newDescription);
    expect(defaultProps.updateDescription).toHaveBeenCalledWith(newDescription);

    expect(tree).toMatchSnapshot();
  });

  it('updates internal state correctly', () => {
    let tree;
    act(() => {
      tree = renderer.create(<Description {...defaultProps}/>);
    });

    const inlineEditable = tree.root.findByType(InlineEditable);
    const input = inlineEditable.props.children;

    // Test internal state updates
    const editedDescription = 'Test description edit';
    act(() => {
      input.props.onChange({target: {value: editedDescription}});
    });

    expect(inlineEditable.props.value).toBe(editedDescription);

    // Test props override internal state updates
    const newProps = {
      ...defaultProps,
      description: 'New description'
    };

    act(() => {
      tree.update(<Description {...newProps}/>);
    });

    expect(inlineEditable.props.value).toBe(newProps.description);
  });
});
