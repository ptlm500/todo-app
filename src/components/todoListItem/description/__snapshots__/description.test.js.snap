// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Description renders a Description 1`] = `
<div
  className="sc-AxirZ ixalE test-class inline-editable"
>
  <div
    className="inline-editable__edit-trigger"
    onClick={[Function]}
  >
    <span
      className="inline-editable__value"
    >
      Test description
    </span>
  </div>
</div>
`;
